### ISSUE TITLE
Update to MathJax 3

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#7218]
MathJax 3 was released a few days ago. Perhaps we should upgrade to it in 2.0.

PR for our external plugin is at https://github.com/jupyterlab/jupyter-renderers/pull/213

**Comment by jasongrout** [comment]
Perusing the [MathJax 3 upgrade notes](http://docs.mathjax.org/en/latest/upgrading/v2.html), I think we should hold off until at least the safe extension is ported to v3, as well as the linebreaking.

**Comment by jasongrout** [comment]
It looks like the safe extension is now available in MathJax 3.1: https://docs.mathjax.org/en/v3.1-latest/upgrading/whats-new-3.1.html#new-safe-extension

**Comment by randolf-scholz** [comment]
It would be really great if MathJax3 support also includes extensions, especially the physics package http://docs.mathjax.org/en/latest/input/tex/extensions/physics.html is an incredible helper when documenting physics/math heavy notebooks.

**Comment by jasongrout** [comment]
When this has come up before, I think we decided to provide the `require` package, then you can load it yourself in your mathjax. From http://docs.mathjax.org/en/latest/input/tex/extensions/physics.html:

> Alternatively, use \require{physics} in a TeX expression to load it dynamically from within the math on the page, if the require package is loaded.

Does `\require{physics}` work for you?

**Comment by randolf-scholz** [comment]
@jasongrout Unfortunately not, I tried using it after installing the [@jupyterlab/mathjax3-extension](https://github.com/jupyterlab/jupyter-renderers/tree/master/packages/mathjax3-extension), but it did nothing. Another user already created an issue here: github.com/jupyterlab/jupyter-renderers/issues/229

**Comment by jasongrout** [comment]
Thanks. I looked into a bit, and put some of my findings over on that issue.

**Comment by jasongrout** [comment]
Line breaking in MathJax 3 is discussed at https://github.com/mathjax/MathJax/issues/2312

**Comment by akhmerov** [comment]
Following this comment: https://github.com/mathjax/MathJax/issues/2312#issuecomment-933057154 I believe what is meant by the lack of line breaking support will not impact most users, and technically doesn't even correspond to valid latex. Is that perhaps a reason to not wait for further improvements?

**Comment by mgeier** [comment]
Is there anything still blocking this?

**Comment by krassowski** [comment]
In another thread we discussed jumping straight to 4.0 but I would be happy to review a PR for v3. Ideally it would come with a PR to jupyterlab-renderers adding v2 as a fallback.

**Comment by mgeier** [comment]
> In another thread

Can you please provide a link?

> we discussed jumping straight to 4.0

But when is that supposed to come out?

The milestone has no due date and is 1% complete: https://github.com/mathjax/MathJax/milestone/29

> but I would be happy to review a PR for v3.

I can't provide a PR, sorry.

But I'm wondering why this doesn't seem to have higher priority within the team. Isn't this an essential feature of JupyterLab?
The user experience with MathJax v3 is just much better than v2. It's much faster, it doesn't move around that much during loading and it also looks better.

**Comment by krassowski** [comment]
> Can you please provide a link?

https://github.com/jupyterlab/jupyterlab/issues/9757#issuecomment-12738067320. Strange that GH does not show it as linked above.

> But when is that supposed to come out?

Good question, I'm afraid wrong place to ask it if you want an answer.

> The milestone has no due date and is 1% complete: https://github.com/mathjax/MathJax/milestone/29

This one says 58%: https://github.com/mathjax/MathJax-src/milestone/10. I have no insight, but many projects (including JupyterLab assign milestones on "if there is PR and time to test it we are happy to review and merge it for this release" basis, and then bump milestones as needed just before release).

> why this doesn't seem to have higher priority within the team

The premise may be wrong. But for one, no one asked here in over a year. Possibly because users can now pip-install [`jupyterlab-mathjax3`](https://pypi.org/project/jupyterlab-mathjax3/).

> The user experience with MathJax v3 is just much better than v2. It's much faster, it doesn't move around that much during loading and it also looks better.

Yes it does. And [it is faster](https://github.com/jupyterlab/jupyterlab/issues/9757#issuecomment-1329520925). Again, I can help with PR. The code is in jupyterlab-mathjax3, it is "just" getting the PRs over there in, making a PR here and there and testing comprehensively.

**Comment by mgeier** [comment]
> The premise may be wrong. But for one, no one asked here in over a year.

I assume that people simply don't know that they are getting a sub-par experience.

Most people might not be familiar with MathJax, and even less with the fact that there might be a much better version available out there.

This is really not something you should wait for more users to complain about.

You should simply do it.

> Possibly because users can now pip-install [`jupyterlab-mathjax3`](https://pypi.org/project/jupyterlab-mathjax3/).

I don't know if that's it.

Of the 1.5 millon JupyterLab downloads a month, about 400 have also downloaded `jupyterlab-mathjax3`. I would say that most users are missing out on MathJax 3.

https://pypistats.org/packages/jupyterlab-mathjax3
https://pypistats.org/packages/jupyterlab
https://pepy.tech/project/jupyterlab-mathjax3
https://pepy.tech/project/jupyterlab

> Again, I can help with PR.

Again, I can't provide a PR, sorry.

**Comment by dlqqq** [comment]
@krassowski Hey Mike, I'm happy to take ownership of this issue and work towards upgrading us to MathJax 3. We have had some AWS customers complain about performance, so optimizing JL 4 performance is a high priority for us. A few questions:

1. Can we drop the version number from the extension name, so we don't need to go through the process of creating a new package again when MathJax 4 is released? IOW, the new path of the extension in this repo would be truncated to `packages/mathjax`, and it would be published on NPM as `@jupyterlab/mathjax`. I checked NPM, and there's no existing package registered under that alias.

2. Can we also drop the "dual-package" pattern we're following for labextensions? Most packages under the JL monorepo are actually two packages:
   - one "main" package containing all the relevant source code
   
   - one "plugin" package that just imports the code from the main package, wraps it in a JupyterFrontEndPlugin, and default exports the plugin.

   I don't really see a purpose for this convention; I suspect it is a legacy pattern that existed before [declaratively disabling extensions](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#disabling-other-extensions) was possible. We developed Jupyter Scheduler without following this pattern, and there doesn't appear to be any negative side-effects of doing so. I believe that if it's possible to move away from this repetitive pattern, we should do so starting now.

3. Finally, after we upgrade to Mathjax 3 in JL 4, what should we do with the existing [`mathjax3-extension`](https://github.com/jupyterlab/jupyter-renderers/tree/master/packages/mathjax3-extension) package? I suspect we would want to include a notice that this package can only be installed for JL 3, lower the version ceiling to exclude JL >=4.0, and then proceed to deprecate the package when JL 5 comes out and JL 3 gets deprecated.

**Comment by jtpio** [comment]
> 2\. Can we also drop the "dual-package" pattern we're following for labextensions? Most packages under the JL monorepo are actually two packages:

Probably this could be discussed in a separate issue. For now it would be great to continue with the current pattern to avoid confusion.

I remember looking at the PR that implemented that change of splitting the plugins into the `-extension` packages a couple of years. Will post it here if I can find it as it might have more context about this decision.

Probably this was done so the base package could be used in arbitrary applications, for example: https://github.com/jupyterlab/jupyterlab/tree/master/examples/cell. So those who would like to reuse components and attach them to the DOM directly don't need to pull the actual plugins that require a `JupyterFrontEnd` application (for example Notebook 7: https://github.com/jupyter/notebook).

**Comment by krassowski** [comment]
Thank you for offering your help here @dlqqq, much appreciated!

> Can we drop the version number from the extension name, so we don't need to go through the process of creating a new package again when MathJax 4 is released?

I think we could, not sure if we should. The decision to postfix the third-party renderer names with version predates me, but I think it makes sense if a renderer has multiple versions which can be used simultaneously. I don't think it is possible to have multiple versions of MathJax at the same time so I think it could be fine to drop the version (but then, publishing new package is just changing the name, I don't think there is a substantial overhead to it).

I would suggest to bring it up tomorrow at the team meeting. I can do this if you cannot attend.

> Can we also drop the "dual-package" pattern we're following for labextensions? Most packages under the JL monorepo are actually two packages:

Adding to @jtpio's explanation, `-extension` packages nicely separate dependencies reducing what downstream needs to depend on to make use of the "main" packages. As an extension developer I find this useful.

Our documentation mentions this patter briefly [here](https://jupyterlab.readthedocs.io/en/stable/developer/repo.html#npm-packages-packages). Maybe we should expand it to clarify what is the purpose (downstream benefits) of doing so and when exceptions are allowed. For example, `vega5-extension` does not have a corresponding `vega5` core package as it does is just a MIME renderer and does not have any special code (other than third-party implementation) that we would want to expose to downstream without extra dependencies.

> Finally, after we upgrade to Mathjax 3 in JL 4, what should we do with the existing [mathjax3-extension](https://github.com/jupyterlab/jupyter-renderers/tree/master/packages/mathjax3-extension) package?

This question does not exist if we just keep doing what we are doing and move `mathjax3-extension` to core. 

> I suspect we would want to include a notice that this package can only be installed for JL 3, lower the version ceiling to exclude JL >=4.0 

Am I right to say that if we just move `mathjax3-extension` to this repository then we don't need to do anything? If I recall correctly, ut already "just works" in both 3.x and 4.x.

**Comment by dlqqq** [comment]
@krassowski 

> I would suggest to bring it up tomorrow at the team meeting. I can do this if you cannot attend.

Don't worry, I can attend tomorrow to discuss this.

> Am I right to say that if we just move mathjax3-extension to this repository then we don't need to do anything? If I recall correctly, it already "just works" in both 3.x and 4.x.

Well, if we were to go along with my suggestion and create a separate `mathjax-extension` package (without the version number in the name), then we would need to provide a deprecation notice for `mathjax3-extension`, as that would likely be abandoned in favor of the former. But this is unnecessary if we decide to keep the `mathjax3-extension` as the package name.

In summary, both strategies result in MathJax 3 support. My suggestion provides a more stable naming convention but forces us to deprecate an existing package, the other one "goes with the flow" but forces us to repeat this process for every new major version of MathJax.

**Comment by jasongrout** [comment]
For context,  [MathJax 4 alpha 1](https://github.com/mathjax/MathJax/releases/tag/4.0.0-alpha.1) has been released, and brings a feature that was our last compatibility issue with MathJax 3, namely automatic line breaking.

If someone depended on this automatic linebreaking, they likely will want to stay with MathJax 2 or upgrade directly to MathJax 4. How do we have a package to provide the MathJax 4 alpha if a user wants to use it? If we strip out the Mathjax version name from our package, then perhaps we can publish an alpha version of our new MathJax package that uses MathJax 4 that users can install?

**Comment by williamstein** [comment]
I'm thrilled that Mathjax v4 alpha finally landed linebreak support!  That said, I think it might be helpful to answer this question that was asked above in 2021:

> Following this comment: [mathjax/MathJax#2312 (comment)](https://github.com/mathjax/MathJax/issues/2312#issuecomment-933057154) I believe what is meant by the lack of line breaking support will not impact most users, and technically doesn't even correspond to valid latex. Is that perhaps a reason to not wait for further improvements?

Here's a screenshot comparing standard valid latex and Mathjax v3 (from [their website demo](https://www.mathjax.org/#demo)):

<img width="1557" alt="image" src="https://user-images.githubusercontent.com/1276278/214909674-758896d6-db1d-4a0b-a2da-7f66c7396479.png">

I'm putting the following all on one line, like Mathjax v3 formula output is, to emphasize that lack of word wrap is a very real problem, as you try to read it:
```txt
The lack of line breaking support in Mathjax v3 could impact many users in that it will change what they see.  Also, it's not totally fair to say "technically doesn't even correspond to valid latex" as illustrated in the screenshot above.   This is partly an accessibility issue, since somebody with very good vision and a large monitor might not be impacted, whereas somebody who is visually impaired with a higher zoom might be more impacted.
```

---

Footnote: here's the input for my little example:
```tex
\documentclass{article}
\begin{document}
Consider $f = x^{20} + 20x^{19} + 190x^{18} + 1140x^{17} + 4845x^{16} + 15504x^{15} + 38760x^{14} + 77520x^{13} + 125970x^{12} + 167960x^{11} + 184756x^{10} + 167960x^{9} + 125970x^{8} + 77520x^{7} + 38760x^{6} + 15504x^{5} + 4845x^{4} + 1140x^{3} + 190x^{2} + 20x + 1$.
\end{document}
```
